### PR TITLE
[MM-41288] Add Setup Wizard

### PR DESCRIPTION
**Pull request by hanzei**
#### Summary
This is a draft of the MVP Setup Wizard

### TODO
- [x] Add telemetry for completing the whole wizard
- [x] Add telemetry stats (https://mattermost.atlassian.net/browse/MM-41289)

#### Ticket Link

https://mattermost.atlassian.net/browse/MM-40924
https://mattermost.atlassian.net/browse/MM-41288
https://mattermost.atlassian.net/browse/MM-41289

### CODE REVIEW

File: server/plugin/api.go [hunk 1]
```diff
@@ -372,32 +379,48 @@ func (p *Plugin) completeConnectUserToGitHub(c *Context, w http.ResponseWriter,
...
+			message := fmt.Sprintf("#### Welcome to the Mattermost GitHub Plugin!\n"+
+				"You've connected your Mattermost account to [%s](%s) on GitHub. Read about the features of this plugin below:\n\n"+
+				"##### Daily Reminders\n"+
+				"The first time you log in each day, you will get a post right here letting you know what messages you need to read and what pull requests are awaiting your review.\n"+
+				"Turn off reminders with `/github settings reminders off`.\n\n"+
+				"##### Notifications\n"+
+				"When someone mentions you, requests your review, comments on or modifies one of your pull requests/issues, or assigns you, you'll get a post here about it.\n"+
+				"Turn off notifications with `/github settings notifications off`.\n\n"+
+				"##### Sidebar Buttons\n"+
+				"Check out the buttons in the left-hand sidebar of Mattermost.\n"+
```
Comment (aaronrothschild) [inline server/plugin/api.go:411]: Can we add a screenshot of the "notification bar" here?

Comment (aaronrothschild) [inline server/plugin/api.go:411]:
```suggestion
				"Check out the buttons in the left-hand sidebar of Mattermost.\n"+
				"It shows your Open PRs, PR that are awaiting your review, issues assigned to you, and all your unread messages you have in Github. \n"+
```

File: server/plugin/command.go [hunk 2]
```diff
@@ -532,8 +585,7 @@ func (p *Plugin) ExecuteCommand(c *plugin.Context, args *model.CommandArgs) (*mo
 			text = "Error checking user's permissions"
 			p.API.LogWarn(text, "err", err.Error())
 		case isSysAdmin:
-			githubPluginURL := *p.API.GetConfig().ServiceSettings.SiteURL + "/admin_console/plugins/plugin_github"
-			text = fmt.Sprintf("Before using this plugin, you will need to configure it by filling out the settings in the system console [here](%s). You can learn more about the setup process [here](%s).", githubPluginURL, "https://github.com/mattermost/mattermost-plugin-github#step-3-configure-the-plugin-in-mattermost")
+			text = "Before using this plugin, you will need to configure it by running `/github get-started`"
```
Comment (aaronrothschild) [inline server/plugin/command.go:588]: should we also say "or, set it up in the system console [here](link to plugin console page)"?  I'm 0/5 on driving the admins through the wizard by omitting this instruction.  @Rina-dsg should we include the alternative way?

**Review by aaronrothschild — APPROVED** [review]
LGTM.  How can I see this deployed on a server to test it out?  cc @DHaussermann

**Comment by hanzei** [comment]
> LGTM. How can I see this deployed on a server to test it out? the plugin isn't installed on the spinwick... cc @DHaussermann

Unfortunately, not right now. I've reached out to the cloud team about this: https://community-daily.mattermost.com/core/pl/qg6u7q4f6jgc9dq3k8ogxaw4mc

File: server/plugin/command.go (hunk 2 above)
Comment (Rina-dsg) [reply server/plugin/command.go]: @aaronrothschild I would push admins to go through the wizard for the first iteration of the wizard in order to test this solution and understand what we can improve.

File: server/plugin/command.go [hunk 3]
```diff
@@ -508,6 +510,43 @@ func (p *Plugin) handleIssue(_ *plugin.Context, args *model.CommandArgs, paramet
...
+	userID := args.UserId
+	isSysAdmin, err := p.isAuthorizedSysAdmin(userID)
+	if err != nil {
+		p.API.LogWarn("Failed to check if user is System Admin", "err", err.Error())
+
+		return "Error checking user's permissions"
+	}
+
+	if !isSysAdmin {
+		return "Only System Admins are allowed to setup the plugin."
```
Comment (cwarnermm) [inline server/plugin/command.go:523]: "setup" here would be 2 words: "set up"

File: server/plugin/command.go [hunk 4]
```diff
@@ -532,8 +585,7 @@ func (p *Plugin) ExecuteCommand(c *plugin.Context, args *model.CommandArgs) (*mo
 			text = "Error checking user's permissions"
 			p.API.LogWarn(text, "err", err.Error())
 		case isSysAdmin:
-			githubPluginURL := *p.API.GetConfig().ServiceSettings.SiteURL + "/admin_console/plugins/plugin_github"
-			text = fmt.Sprintf("Before using this plugin, you will need to configure it by filling out the settings in the system console [here](%s). You can learn more about the setup process [here](%s).", githubPluginURL, "https://github.com/mattermost/mattermost-plugin-github#step-3-configure-the-plugin-in-mattermost")
+			text = "Before using this plugin, you will need to configure it by running `/github setup`"
```
Comment (cwarnermm) [inline server/plugin/command.go:588]:
```suggestion
			text = "Before using this plugin, you'll need to configure it by running `/github setup`"
```

**Review by cwarnermm — COMMENTED** [review]
Provided non-blocking feedback inline for a couple of display text elements.

File: server/plugin/command.go (hunk 4 above)
Comment (hanzei) [reply server/plugin/command.go]: Fixed in 4db6b6a

File: server/plugin/command.go (hunk 3 above)
Comment (hanzei) [reply server/plugin/command.go]: Fixed in 4db6b6a

**Comment by mickmister** [comment]
@aaronrothschild @cwarnermm The majority of the content changes that need to be reviewed are in `flows.go`. I just wanted to point this out because I assume the file wasn't reviewed. You can view the contents by clicking "Load diff". This happens when the file has a large number of changes.




Also for changes like the one in the image below, where the content in the diff looks mostly the same, but the GitHub diff reports many changes, this is likely due to whitespace-only changes.




In order to properly review this change (to see specifically what has changed in a given pull request), you can "hide" the whitespace changes. This way only text content that actually changed can be reviewed



After applying the "Hide whitespace" setting, you can see the diff to review is only about 2 lines, though it originally looked like entire paragraphs had changed. This makes our reviews more precise as it's scoped to only the content changes introduced by the PR.




Just pointing these items out so we get the most out of the time we spend reviewing pull requests 🙂

@cwarnermm Maybe we can document these tips in some PR review guide we may have? I'm not sure if something like that exists

**Review by mickmister — CHANGES_REQUESTED** [review]
Awesome work @hanzei! Thanks for going forward with adding telemetry etc.

I have a few questions and some suggestions for minor improvements. Overall looks good :tada:

**Comment by DHaussermann** [comment]
@hanzei some follow up questions on this...
1. Is there a future commit coming that will ask me about using the Chimera Proxy App? I ran this on a cloud server where the proxy URL should be in the config and did not see a prompt to use it.  
2. Can we communicate about the other 2 setup steps `webhook` and `announcement` once the admin is connected? 
1/5 - Is there a way to send another DM once the admin is connected telling them there are new option for this in the setup command? we show some info about a step 3. here at the beginning but we don't prompt the admin to do it or show the explicit commands.
image
3. Maybe add `(if prompted)` to step 5 of oAuth as the user does not always get prompted for example if they have just logged in to make the app)
image

**note**: These 👇  are higher level not specific to GitHub
4. Move the red `*` to be on the same line (I think this was also brought up in the channel discussion)
image
5. Can we provide feedback **if** the user running through setup is overwriting oAuth app values that are already in the config? (maybe a nice to have for the future)
My concern here would be that the setup is passed to a user who does not action it for several days and by the time they start the setup was already done by someone else.
6. Can we remove Bots and the current user form the single select to make this cleaner?

File: server/plugin/flows.go [hunk 5]
```diff
@@ -0,0 +1,763 @@
...
+					Title:       "Create Webhook",
+					SubmitLabel: "Create",
+					Elements: []model.DialogElement{
+						{
+
+							DisplayName: "", // TODO: This will still show a *
+							Name:        "repo_org",
+							Type:        "text",
+							SubType:     "text",
+							Placeholder: "Repository or organization name, e.g. mattermost/mattermost-server",
```
Comment (cwarnermm) [inline server/plugin/flows.go:524]:
```suggestion
							Placeholder: "Enter GitHub repository or organization name",
```

File: server/plugin/flows.go [hunk 6]
```diff
@@ -0,0 +1,763 @@
...
+							HelpText:    "Specify repository or organization for which you want to create a webhook",
```
Comment (cwarnermm) [inline server/plugin/flows.go:525]:
```suggestion
							HelpText:    "Specify the GitHub repository or organization to connect to Mattermost. For example, mattermost/mattermost-server.",
```

File: server/plugin/flows.go [hunk 7]
```diff
@@ -0,0 +1,763 @@
...
+	if repo == "" {
+		fullName = org
+		hook, resp, err = client.Organizations.CreateHook(ctx, org, hook)
+	} else {
+		fullName = org + "/" + repo
+		hook, resp, err = client.Repositories.CreateHook(ctx, org, repo, hook)
+	}
+
+	if resp.StatusCode == http.StatusNotFound {
+		return 0, nil, fmt.Sprintf("It seems like you don't have access %s. Please ask an administrator of that repository to run /github setup webhook for you.", fullName), nil
```
Comment (cwarnermm) [inline server/plugin/flows.go:610]:
```suggestion
		return 0, nil, fmt.Sprintf("It seems like you don't have access %s. Ask a system admin of that repository to run `/github setup` webhook for you.", fullName), nil
```

**Comment by cwarnermm** [comment]
Thank you, @mickmister, for directing @aaronrothschild and I to the areas of the code we'd be most likely to want to review. I've completed an initial code review pass and identified text updates inline. I'll complete my next pass using the preview environment to review all of the wizard text contextually.

**Review by levb — APPROVED** [review]
I'd like to see the webhook handling code to be HA-aware, otherwise 1/5 merge this and refine (the copy) separately.

**Comment by Rina-dsg** [comment]
@hanzei Popups are missing the titles for input fields.
Screenshot 2022-02-01 at 15 36 05

Example
Screenshot 2022-02-01 at 15 37 30

File: server/plugin/flows.go (hunk 7 above)
Comment (hanzei) [reply server/plugin/flows.go:619]: No support for Markdown in the error message :confused:

File: server/plugin/command.go [hunk 8]
```diff
@@ -716,12 +771,18 @@ func getAutocompleteData(config *Configuration) *model.AutocompleteData {
...
+	issueCreate.AddTextArgument("Title for the GitHub issue", "[title]", "")
 	issue.AddCommand(issueCreate)
 
 	github.AddCommand(issue)
 
+	setup := model.NewAutocompleteData("setup", "[command]", "Available commands: oauth, webhook, announcement")
+	setup.AddCommand(model.NewAutocompleteData("oauth", "", "Set up the OAuth2 Application in GitHub"))
+	setup.AddCommand(model.NewAutocompleteData("webhook", "", "Create a webhook from GitHub to Mattermost"))
+	setup.AddCommand(model.NewAutocompleteData("announcement", "", "Announce to your team that they can use GitHub integration"))
+	github.AddCommand(setup)
```
[the remark this answers is not in the document]
Comment (hanzei) [reply server/plugin/command.go]: Good suggestion, fixed in b07a43f.

File: server/plugin/configuration.go [hunk 9]
```diff
@@ -34,6 +37,73 @@ type Configuration struct {
...
+		if c.WebhookSecret == "" {
+			secret, err := generateSecret()
+			if err != nil {
+				return false, err
+			}
+
+			c.WebhookSecret = secret
+			changed = true
+		}
+	}
```
[the remark this answers is not in the document]
Comment (hanzei) [reply server/plugin/configuration.go:94]: Don't know, fixed

File: server/plugin/configuration.go [hunk 10]
```diff
@@ -34,6 +37,73 @@ type Configuration struct {
...
+	return changed, nil
+}
+
+func (c *Configuration) getBaseURL() string {
+	if c.EnterpriseBaseURL != "" {
+		return c.EnterpriseBaseURL
+	}
+
+	return "https://github.com/"
+}
```
[the remark this answers is not in the document]
Comment (hanzei) [reply server/plugin/configuration.go:103]: No, we didn't. Fixed that in 3cb40b6418ae467bbf8f733b54bc40b97ff7fc63.

File: server/plugin/flows.go [hunk 11]
```diff
@@ -0,0 +1,763 @@
...
+							Name:        "client_id",
+							Type:        "text",
+							SubType:     "text",
+							Placeholder: "GitHub OAuth Client ID",
+						},
+						{
+							DisplayName: "", // TODO: This will still show a *
+							Name:        "client_secret",
+							Type:        "text",
+							SubType:     "text",
```
[the remark this answers is not in the document]
Comment (hanzei) [reply server/plugin/flows.go]: Good suggestion, applied in cfb08ac

File: server/plugin/flows.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (hanzei) [reply server/plugin/flows.go]: 7551e9c

File: server/plugin/flows.go [hunk 12]
```diff
@@ -0,0 +1,763 @@
...
+			},
+		}).
+		WithButton(steps.Button{
+			Name:    "Cancel setup",
+			Style:   steps.Danger,
+			OnClick: fm.cancelFlow,
+		}).
+		Build()
+
+	confirmationStep := steps.NewEmptyStep("success", "Success! :tada: You've successfully set up your Mattermost GitHub integration! ", "Use `/github subscriptions add` to subscribe any Mattermost channel to your GitHub repository. [Learn more](https://example.org)")
```
[the remark this answers is not in the document]
Comment (hanzei) [reply server/plugin/flows.go:557]: @cwarnermm To what should be linked here?

File: server/plugin/flows.go [hunk 13]
```diff
@@ -0,0 +1,763 @@
...
+	}
+
+	return steps
+}
+
+func (fm *FlowManager) submitHandoverSelection(userID string, submission map[string]interface{}) (int, *steps.Attachment, string, map[string]string) {
+	aiderIDRaw, ok := submission["aider"]
+	if !ok {
+		return 0, nil, "aider missing", nil
+	}
```
[the remark this answers is not in the document]
Comment (hanzei) [reply server/plugin/flows.go:354]: `other_admin` is misleading to be, as the other person is exactly **not** an system admin. I also don't like `aider`, so I'm open for other suggestions.

File: server/plugin/api.go (hunk 1 above)
Comment (hanzei) [reply server/plugin/api.go]: The message already is quite long. 2/5 let's not add images there without a mayor re-write.

**Comment by hanzei** [comment]
@Rina-dsg @cwarnermm All of your feedback has been implemented. Please take another look.

**Comment by hanzei** [comment]
@mickmister Your feedback has been addressed. Please take another look.

@levb I've added HA-aware handling of webhook and oauth events in 8084779409923c98668f0d5a9f567388a15a7653. Please take another look.

**Comment by hanzei** [comment]
@DHaussermann We already talked about your points offline, here is the summary:

1. In Cloud, Chimera should be used by default. I will talk with the cloud team about this.
2. Please re-test with the latest version.
3. :+1: Added that
4. That requires changes to the framework, won't do.
5. This is a TODO for me to implement
6. That requires changes to the framework, won't do.

File: server/plugin/flows.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (hanzei) [reply server/plugin/flows.go]: I went with `delegate`.

**Comment by Rina-dsg** [comment]
@hanzei a small mistypo in the heading, instead of Oauth should be OAuth
Screenshot 2022-02-03 at 15 05 04

- let's kill `to` in the heading here
Screenshot 2022-02-03 at 15 04 03

- It looks like the link is broked in the first step
Screenshot 2022-02-03 at 15 17 35

**Comment by hanzei** [comment]
@Rina-dsg Thanks for the review, fixed these issues

**Comment by hanzei** [comment]
@Rina-dsg I can't reproduce the issue. Maybe it was caused by a misconfigured state? Was there an enterprise URL set?

**Comment by Rina-dsg** [comment]
@hanzei not sure, it looks fine now.

**Comment by hanzei** [comment]
@levb @mickmister Heads up that I updated the PR to include the changes from @levb in https://github.com/mattermost/mattermost-plugin-api/pull/111

**Comment by DHaussermann** [comment]
Hi @hanzei, I have done another round of testing and have some points to discuss..
1. There is a JS error I'm seeing. I have no concrete repro steps. Possibly occurs on connect/disconnect. 
`utils.js`
image
image
2. Should we hid the `setup` command from the auto-complete
3. When I run the Enterprise branch of the setup - I see a strange issue where the setup has written the values to the config however, the UI will not show the values. Even after a server restart, the fields still look blank. If it's relevant I have `MM-41200_OnSendDailyTelemetry_hook` deployed for server and `use-case-onboarding-wizard` for webapp.  
 Any thoughts?
image
4. Similar to 3. - `EncryptionKey` and` WebhookSecret` get written to the config file on start-up but are not visible in the UI.
5. Chimera is enabled, the proxy config does not seem to work. The connection link points to `https://dkh-local.ngrok.io/plugins/github/oauth/connect` and I then ge redirected to `https://dkh-local.ngrok.io/v1/github/plugin-github/oauth/authorize?access_type=offline&client_id=placeholder&redirect_uri=https%3A%2F%2Fdkh-local.ngrok.io%2Fplugins%2Fgithub%2Foauth%2Fcomplete&response_type=code&scope=public_repo+notifications+read%3Aorg+admin%3Aorg_hook&state=d8brmbfrgbf67b4`
Should `client_id=placeholder` be parameterized?
image
image
6. Even with Server and changes in the `MM-41200_OnSendDailyTelemetry_hook` branch deployed for server, I must run the setup command manually after enabling the plugin. Is this expected?

@cwarnermm A couple more points about the wording...
A. Step 3. Should read "Register Application" to match GitHub UI instead of Submit
image
B. 0/5 on this but maybe the wording to open the dialogue is a bit more clear if you use the same wording as we do for Zoom
image
This could read `Click the Continue button below to open a dialog to enter the **GitHub OAuth Client ID** and **GitHub OAuth Client Secret**.`

**Comment by hanzei** [comment]
Thanks for the review @DHaussermann :+1: 

1. Reproduces on `master`. Filled a separate ticket: https://github.com/mattermost/mattermost-plugin-github/issues/530
2. Why should that be done?
3. I have not seen that issue before. Did you test in HA?
4. See 3.
5. Discussed offline: The chimera URL seems incorrect
6. Yes. The setup is only executed if the plugin is installed via the Onboarding flow. Please note that `MM-40935_OnInstall` must be used for that.

**Comment by DHaussermann** [comment]
@hanzei U have investigated issues 3. and 4. above. 
Local testing would suggest that this is somehow a change in the GitHub plugin? 

**Case A**:
Server and webapp set to `5.39` branch and this branch of GitHub plugin shows the same behavior.

**Case B**
Server and webapp set to `5.39` branch and GitHub production `v2.0.1` then no longer shows the key duplication

 **Case C**
Server and webapp set to `current` branch and GitHub production `v2.0.1` alson o longer show the duplicatyoon

**Comment by DHaussermann** [comment]
@hanzei I have updated the post above ☝️ :  Let me know if there is anything else I can check. tis seems reasonably conclusive.

**Review by catalintomai — APPROVED** [review]

**Comment by hanzei** [comment]
@DHaussermann Thank you very for the investigation :+1: I've fixed the configuration handling in this plugin. PTAL

**Comment by DHaussermann** [comment]
@hanzei The config issue had been resolved. Huge thanks!

I do see a couple more issues with the latest commit deployed to a cloud server. Let mw know if anything can be done about 1. before you merge this. 

1. I don't think I'm seeing the same behavior as you for environments where I want to use the Chimera proxy app. I have seen the other branch work before I think but for some reason, now it is showing me the flow for creating the oAuth app. Any thoughts on what may be happening here?

Steps: 
- Created a cloud server (uses miniHA if that's related)
- Uploaded the build form this latest commit
- Enabled the plugin 
- Enabled Chimera proxy app and saved (Connect options are visible which makes sense)
- Run Setup command 
**Observed**: I am in the branch with oAuth app creation and can't walk through the wizzard to do step 2. and 3. without making an app
image

2. Intermittently, I see an issue where the dialogue is showing credential already that are cached in the browser. Seems like maybe a framework level issue. I have seen it with Zoom flow as well but then, could not repro. Any thoughts?
image

**Comment by cwarnermm** [comment]
@DHaussermann - I agree with both of your text suggestions for the wizard. Would you be open to applying both updates in this PR?

**Comment by DHaussermann** [comment]
@hanzei For issue 1. This seems worth investigating. That said it may not be applicable with the default config when the license is `cloud` type. Therefore likely would **not** reproduce on proper config. (For reference -  in the test above I cannot apply a cloud license AND upload a plugin manually) 

Here were my local steps:
- Remove the GitHub plugin and records of it in the config.json
- Set a token value for `ChimeraOAuthProxyURL` in the config to allow plugin start-up with proxy
- Upload plugin  from this branch (do not enable)
- Apply a `cloud` license 
- Enable the plugin
**Observed**: 
A. Plugin starts up with use of proxy app `true`
B. Running `/github setup` will show correct 2-step branch on the 1st attempt

Given this, I would suggest we merge this now a begin the release testing process. If there are further changes to cherry-pick in, they would not affect the vast majority of release testing.

**Comment by DHaussermann** [comment]
Hi @cwarnermm, After discussing this with Ben today, it was decided we should merge this PR to begin the release testing process. We can cherry-pick in some changes while release testing for the the rest of the plugin is ongoing. 

That being said, would it be possible to open a separate PR for A. and B. above depending on how soon Ben can merge this.? I'm not sure what's required for a separate PR or if I can help in any way. Should I open a n new issue in this repo?

**Comment by hanzei** [comment]
@cwarnermm @DHaussermann I've updated the text based on the feedback. No need for another ticket.

File: server/plugin/flows.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (hanzei) [reply server/plugin/flows.go]: Linked it to https://github.com/mattermost/mattermost-plugin-github#slash-commands

**Review by DHaussermann — APPROVED** [review]
Tested and passed
- Work through workflows with Ben
- Al GitHub specific issues have been addressed
- This should now be merged to master branch
- Last reported issue should not affect properly configured servers.

LGTM!

File: server/plugin/api.go [hunk 14]
```diff
@@ -363,41 +419,60 @@ func (p *Plugin) completeConnectUserToGitHub(c *Context, w http.ResponseWriter,
...
-		"* The fourth tracks the number of unread messages you have.\n"+
-		"* The fifth will refresh the numbers.\n\n"+
-		"Click on them!\n\n"+
-		"##### Slash Commands\n"+
-		commandHelp, gitUser.GetLogin(), gitUser.GetHTMLURL())
-
-	p.CreateBotDMPost(state.UserID, message, "custom_git_welcome")
+		c.Log.WithError(err).Warnf("Failed to get current step")
+	}
+
```
Comment (mickmister) [inline server/plugin/api.go:439]: What happens when this user is not an admin? Do we want to log this warning unconditionally here?

File: server/plugin/oauth.go [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+			case sub <- err:
+			default:
+			}
+		}
+	}
+
+	if !fromCluster {
+		ob.sendOAuthCompleteEvent(OAuthCompleteEvent{UserID: userID, Err: err})
+	}
+}
```
Comment (mickmister) [inline server/plugin/oauth.go:75]: I'm not sure I understand what's going on here. Why do the subscribed channels not need to know what the user id is for this event?

File: server/plugin/api.go (hunk 14 above)
Comment (hanzei) [reply server/plugin/api.go:439]: `GetCurrentStep()` should return `"", nil` where is no flow running. At least this is show the library used to work. @levb Would you mind changing the method in your PR to behave that way?

File: server/plugin/oauth.go (hunk 15 above)
Comment (hanzei) [reply server/plugin/oauth.go:75]: Good catch :+1: Fixed in [`21ad0be` (#541)](https://github.com/mattermost/mattermost-plugin-github/pull/541/commits/21ad0be7d78cb6708e809e0690f39b213dcb0142), which is in https://github.com/mattermost/mattermost-plugin-github/pull/541.

File: server/plugin/api.go (hunk 14 above)
Comment (hanzei) [reply server/plugin/api.go:439]: https://github.com/mattermost/mattermost-plugin-api/issues/113

**Pull request merged**